### PR TITLE
Update README.md

### PR DESCRIPTION
**Pull request by falseywinchnet**
add a simple implementation

### CODE REVIEW

**Comment by falseywinchnet** [comment]
## Simple Implementation
If you want a super simple, easy to use drop-in form of goldfish loss, here is how you can do it:
```python
import random

def _goldfish_cluster_counts(seq_len: int, drop_frac: float) -> dict:
    """
    Decide how many clusters of each length to use so that each cluster length
    gets an equal token budget. For sizes s=1..K: count_s is ~constant/s.
    We choose K ~= floor(sqrt(N)) so the largest size still has at least 1 cluster.
    """
    N = max(0, int(round(seq_len * drop_frac)))
    if N == 0:
        return {}

    K = max(1, int(np.sqrt(N)))  # largest cluster size
    per_size_budget = N // K     # tokens per size bucket

    counts = {s: (per_size_budget // s) for s in range(1, K + 1)}
    dropped = sum(s * c for s, c in counts.items())
    rem = N - dropped

    # Distribute remainder by giving +1 cluster to random sizes where it fits.
    # This keeps the per-size token totals close while staying simple.
    sizes = list(range(1, K + 1))
    while rem > 0:
        random.shuffle(sizes)
        progressed = False
        for s in sizes:
            if rem >= s:
                counts[s] += 1
                rem -= s
                progressed = True
                if rem == 0:
                    break
        if not progressed:
            break  # cannot fit any more without exceeding N

    # Prune zero entries
    return {s: c for s, c in counts.items() if c > 0}


def _goldfish_place_clusters(seq_len: int, counts: dict, rng: np.random.Generator) -> np.ndarray:
    """
    Place clusters from largest to smallest. For each placed cluster of size s
    starting at offset o, we:
      - drop indices [o, o+s-1]
      - remove from future *offset candidates* the range [o-(s-1), o+s-1] to avoid overlap
    The guard region is not dropped; it only blocks future placements.
    """
    if seq_len <= 0:
        return np.zeros(0, dtype=bool)

    available_for_start = np.ones(seq_len, dtype=bool)  # candidate starts
    drop = np.zeros(seq_len, dtype=bool)

    # Helper to find valid starts for a given size s
    def valid_starts(s: int) -> np.ndarray:
        # A start i is valid if i..i+s-1 are within bounds and not already dropped or blocked
        # We test availability by ensuring all s tokens are currently available for start
        # Using convolution for speed
        window = np.ones(s, dtype=np.int32)
        # Start availability is only about "available_for_start"; dropped tokens also block starts via available_for_start updates
        avail_int = available_for_start.astype(np.int32)
        sums = np.convolve(avail_int, window, mode='valid')
        return np.flatnonzero(sums == s)

    # Place larger clusters first
    for s in sorted(counts.keys(), reverse=True):
        for _ in range(counts[s]):
            starts = valid_starts(s)
            if starts.size == 0:
                continue
            o = int(rng.choice(starts))

            # Drop the s tokens
            drop[o:o + s] = True

            # Block future starts in [o-(s-1), o+s-1]
            left = max(0, o - (s - 1))
            right = min(seq_len, o + s)  # right is exclusive for slicing
            available_for_start[left:right] = False

    # If we still undershot the target due to placement conflicts, top up with singles
    need = int(sum(s * c for s, c in counts.items())) - int(drop.sum())
    if need > 0:
        candidates = np.flatnonzero(~drop)
        if candidates.size > 0:
            picks = rng.choice(candidates, size=min(need, candidates.size), replace=False)
            drop[picks] = True

    return drop


def build_goldfish_keep_mask(T: int, drop_frac: float = 1/np.e, batch_size: int = 1, device=None, seed: int | None = None) -> torch.Tensor:
    """
    Returns a mask of shape [B, T] where 1.0 means keep, 0.0 means drop.
    About 36.8% of positions are dropped per sequence by default.
    """
    rng = np.random.default_rng(seed)
    masks = []
    for _ in range(batch_size):
        counts = _goldfish_cluster_counts(T, drop_frac)
        drop = _goldfish_place_clusters(T, counts, rng)
        keep = (~drop).astype(np.float32)
        masks.append(torch.from_numpy(keep))

    mask = torch.stack(masks, dim=0)
    if device is not None:
        mask = mask.to(device)
    return mask

# And in your training block:

optimizer = torch.optim.AdamW(model.parameters(), lr=2e-3)
losses = []
criterion_none = nn.CrossEntropyLoss(reduction='none')  # per-token
goldfish_frac = 1 / np.e  # ~0.367879441

def train_epoch():
    model.train()
    total_loss = 0

    for xb, yb in train_loader:
        xb, yb = xb[0], yb[0]  # unwrap batch dim from DataLoader
        optimizer.zero_grad()

        # Get logits from the model; ignore any internal loss
        logits, _ = model(xb, yb)  # keep API unchanged but we will compute our own loss

        B, T = yb.shape
        # Some models return [B, T, V], others [B, V, T]. Normalize to [B, T, V]:
        if logits.dim() == 3 and logits.shape[1] == T:
            # logits already [B, T, V]
            logits_btV = logits
        elif logits.dim() == 3 and logits.shape[2] == T:
            # logits [B, V, T] -> [B, T, V]
            logits_btV = logits.transpose(1, 2)
        else:
            raise ValueError(f"Unexpected logits shape: {tuple(logits.shape)}")

        # Build keep mask with clustered dropouts
        keep_mask = build_goldfish_keep_mask(
            T=T,
            drop_frac=goldfish_frac,
            batch_size=B,
            device=yb.device,
            seed=None  # or set a seed for reproducibility
        )  # [B, T], float32 in {0,1}

        # Per-token loss
        per_tok = F.cross_entropy(
            logits_btV.reshape(-1, logits_btV.shape[-1]),
            yb.reshape(-1),
            reduction='none'
        ).view(B, T)  # [B, T]

        # Masked mean over remaining tokens
        denom = keep_mask.sum().clamp(min=1.0)
        loss = (per_tok * keep_mask).sum() / denom

        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
        optimizer.step()

        total_loss += float(loss.item())
        losses.append(float(loss.item()))
        print(loss.item())

    return total_loss / len(train_loader)

# === Run Training ===
num_epochs = 10
for epoch in range(1, num_epochs + 1):
    train_loss = train_epoch()
    print(f"Epoch {epoch:2d} | Train loss: {train_loss:.4f}")
```


There are two distinct goals tangled up in goldfish that make my suggestion poor:
(1) make exact regurgitation exponentially unlikely, even after many repeats/near-duplicates;
(2) still let the model learn useful n-grams.

A clustered “goldfish loss” (1/e mask with mixed gap sizes placed randomly per batch) does (2) well, but it’s weaker on (1) under repeats. The paper’s hashed mask was invented precisely for (1).

Random per-occurrence masking means each position is supervised with probability on every repeat. Eventually the model has seen every token at least once, so a duplicated passage can still be memorized.

Content-deterministic hashing drops the same subset of tokens every time a passage appears (based only on the preceding h-gram). That fixes the “never learned” set at rate f regardless of repeats, so the model can’t see a complete chain of supervision for any duplicated span. Your paper stresses this exact point: inconsistent masks across duplicates “leak the entire passage,” hence the localized hash tied to context so “the (h+1)th token is masked in the same way” whenever the same h-gram appears. 

You also contrast against randomized dropout: “it will eventually learn the entire passage” when the same text is seen multiple times with different masks.

However, the caveat is that if you set it too small, the model will, in fact, have issues learning useful information. As h grows, the chance of hitting frequent (h+1)-grams drops (Zipf).

The implication here is that the model MUST and does learn short range patterns based on long range patterns, and your method for blocking this "goldfish loss" is actually blocking it from obtaining certainty about continual patterns by injecting forced "uncertainty" breaks at specific offsets.


So, actually, I think there's a better idea here- why aren't we injecting uncertainty tokens the model cannot memorize?
